File: site/static/js/cc-switch.js
/**
 * cc-switch.js -- browser-side CC Switch configuration builder.
 *
 * This mirrors radar/cc_switch.py so the preview dialog can recompute a
 * configuration when the user switches target app or model, without
 * re-requesting anything. The two implementations must agree; the Python
 * side is authoritative and the test suite asserts the same fixtures
 * against both.
 *
 * The protocol gate is the core rule and is reproduced exactly:
 *
 *   claude   requires anthropic_messages
 *   codex    requires openai_responses
 *   gemini   requires gemini_native
 *   opencode requires openai_chat
 *   openclaw requires openai_chat
 *
 * If the source did not declare any protocol we return `unknown` and
 * produce NO deep link. We do not assume "OpenAI-compatible" from a URL
 * shape, and we do not guess from the model name. `unknown` never becomes
 * a positive or a negative.
 *
 * The generated configuration NEVER contains an API key. The apiKey field
 * is emitted with a null value and a label telling the user to fill it in
 * their own CC Switch.
 */

export const SUPPORTED_VERSION = '3.20.3';
export const SUPPORTED_COMMIT = '06082e1';
export const DEEP_LINK_VERSION = 'v1';
export const MAX_FIELD_LENGTH = 2048;
export const MAX_NAME_LENGTH = 120;

export const APP_PROTOCOL_REQUIREMENTS = {
  claude: ['anthropic_messages'],
  codex: ['openai_responses'],
  gemini: ['gemini_native'],
  opencode: ['openai_chat'],
  openclaw: ['openai_chat'],
};

export const APP_LABELS = {
  claude: 'Claude Code',
  codex: 'Codex CLI',
  gemini: 'Gemini CLI',
  opencode: 'OpenCode',
  openclaw: 'OpenClaw',
};

export const PROTOCOL_LABELS = {
  openai_chat: 'Chat Completions',
  openai_responses: 'Responses API',
  anthropic_messages: 'Anthropic Messages',
  gemini_native: 'Gemini 原生',
};

export const STATUS = {
  READY: 'ready',
  READY_WITHOUT_KEY: 'ready_without_key',
  UNSUPPORTED_PROTOCOL: 'unsupported_protocol',
  UNKNOWN: 'unknown',
  INVALID_ENDPOINT: 'invalid_endpoint',
  NEEDS_CONVERSION: 'needs_conversion',
};

const ENDPOINT_PATTERN = /^https?:\/\/[^\s<>"']+$/i;

/**
 * Normalise an endpoint for the CC Switch link.
 *
 * Mirrors normalise_endpoint() in Python:
 *   * trailing slashes removed
 *   * a duplicated version segment collapsed (…/v1/v1 -> …/v1)
 *   * non-ASCII percent-encoded, because the deep link is a URL
 */
export function normaliseEndpoint(raw) {
  let text = String(raw ?? '').trim();
  if (!text) return '';

  text = text.replace(/\/+$/, '');
  // Collapse a repeated trailing version segment of the same version.
  text = text.replace(/\/(v\d+(?:beta)?)\/\1$/i, '/$1');
  // Collapse the common /v1/v1 case where the first is part of a path.
  text = text.replace(/(\/v\d+(?:beta)?)\/v\d+(?:beta)?$/i, '$1');

  try {
    return encodeURI(decodeURI(text));
  } catch {
    return encodeURI(text);
  }
}

/** Join a path onto an endpoint without duplicating the version segment. */
export function joinEndpoint(endpoint, path) {
  const base = normaliseEndpoint(endpoint);
  const suffix = String(path ?? '').replace(/^\/+/, '');
  if (!base) return suffix;
  if (!suffix) return base;

  const baseVersion = base.match(/\/v\d+(?:beta)?$/i)?.[0]?.toLowerCase() ?? '';
  const pathVersion = suffix.match(/^v\d+(?:beta)?(?=\/|$)/i)?.[0]?.toLowerCase() ?? '';

  if (baseVersion && pathVersion && baseVersion === pathVersion) {
    return `${base}/${suffix.slice(pathVersion.length).replace(/^\/+/, '')}`;
  }
  return `${base}/${suffix}`;
}

export function isValidEndpoint(endpoint) {
  const text = normaliseEndpoint(endpoint);
  return Boolean(text) && text.length <= MAX_FIELD_LENGTH && ENDPOINT_PATTERN.test(text);
}

/** Strip control characters and angle brackets from a user-visible string. */
function safeText(value, limit = MAX_NAME_LENGTH) {
  const text = String(value ?? '')
    .replace(/[\u0000-\u001f\u007f]/g, ' ')
    .replace(/[<>]/g, '')
    .replace(/\s+/g, ' ')
    .trim();
  return text.length > limit ? `${text.slice(0, limit - 1)}…` : text;
}

/**
 * Decide whether the target app can be driven over the declared protocols.
 *
 * Returns a status plus a reason string. The reason is shown to the user
 * verbatim, so it must be explanatory rather than a code.
 */
export function evaluate({ app, protocols = [], baseUrl = '' }) {
  const required = APP_PROTOCOL_REQUIREMENTS[app];
  if (!required) {
    return {
      status: STATUS.UNKNOWN,
      reason: `未知的目标应用：${app}`,
      protocol: null,
      requiresConversion: false,
    };
  }

  if (!isValidEndpoint(baseUrl)) {
    return {
      status: STATUS.INVALID_ENDPOINT,
      reason: '来源没有提供可用的 Base URL，无法生成配置。',
      protocol: null,
      requiresConversion: false,
    };
  }

  const declared = protocols.filter((entry) => entry && entry !== 'unknown');

  if (declared.length === 0) {
    return {
      status: STATUS.UNKNOWN,
      reason: '来源未声明协议，无法判断该服务是否满足目标工具的调用路径。已标记为 unknown，不会生成可用的导入链接。',
      protocol: null,
      requiresConversion: false,
    };
  }

  const usable = declared.find((entry) => required.includes(entry));
  if (usable) {
    return {
      status: STATUS.READY_WITHOUT_KEY,
      reason: '',
      protocol: usable,
      requiresConversion: false,
    };
  }

  // Declared, but not the one this app speaks. Say which one it does speak
  // so the reader can pick a different target app rather than guessing.
  return {
    status: STATUS.UNSUPPORTED_PROTOCOL,
    reason: `来源声明的是 ${declared.map((p) => PROTOCOL_LABELS[p] ?? p).join('、')}，`
      + `而 ${APP_LABELS[app] ?? app} 需要 ${required.map((p) => PROTOCOL_LABELS[p] ?? p).join('、')}。`
      + '两者协议不同，直接导入不会生效，因此不生成导入链接。',
    protocol: declared[0] ?? null,
    requiresConversion: true,
  };
}

/**
 * Build the field list shown in the dialog.
 *
 * Field order matters: the reader fills them top to bottom, so identity
 * comes first, then the endpoint, then the model, then the key.
 */
export function buildFields({ providerName, app, baseUrl, modelId, homepage, notes }) {
  return [
    {
      key: 'name',
      label: 'Provider 名称',
      value: safeText(providerName),
      required: true,
      hint: '在 CC Switch 中显示的供应商名称',
      secret: false,
      copiable: true,
    },
    {
      key: 'app',
      label: '目标应用',
      value: app,
      required: true,
      hint: APP_LABELS[app] ?? app,
      secret: false,
      copiable: false,
    },
    {
      key: 'endpoint',
      label: 'Base URL',
      value: normaliseEndpoint(baseUrl),
      required: true,
      hint: 'API 端点地址；Radar 已检查 /v1 拼接规则，不会重复拼接',
      secret: false,
      copiable: true,
    },
    {
      key: 'model',
      label: '模型 ID',
      value: modelId ? safeText(modelId, 200) : null,
      required: false,
      hint: '精确的上游模型标识',
      secret: false,
      copiable: true,
    },
    {
      key: 'homepage',
      label: '供应商官网',
      value: homepage || null,
      required: false,
      hint: '可选',
      secret: false,
      copiable: true,
    },
    {
      key: 'notes',
      label: '备注',
      value: notes || null,
      required: false,
      hint: '可选；可以写清额度条件或注意事项',
      secret: false,
      copiable: true,
    },
    {
      key: 'apiKey',
      label: 'API Key',
      value: null,
      required: true,
      hint: '请在你自己的 CC Switch 中填写。Radar 不收集、不传输、不记录任何上游密钥。',
      secret: true,
      copiable: false,
    },
  ];
}

/**
 * Assemble the ccswitch:// deep link.
 *
 * Params are encoded with encodeURIComponent so a non-ASCII provider name
 * or a percent sign inside a URL round-trips. `apiKey` is deliberately
 * absent: the specification allows it, and putting a key in a URL is
 * exactly the leak the brief forbids.
 */
export function buildDeepLink({ app, providerName, baseUrl, modelId, homepage, notes }) {
  const params = new URLSearchParams();

  params.set('resource', 'provider');
  params.set('app', app);
  params.set('name', safeText(providerName));

  const endpoint = normaliseEndpoint(baseUrl);
  if (endpoint) params.set('endpoint', endpoint);

  if (modelId) params.set('model', safeText(modelId, 200));
  if (homepage) params.set('homepage', homepage);
  if (notes) params.set('notes', safeText(notes, 500));

  // Ask CC Switch to leave the entry enabled but keyless until the user
  // supplies a key, rather than importing a broken disabled entry.
  params.set('enabled', 'true');

  return `ccswitch://v${DEEP_LINK_VERSION.replace(/^v/, '')}/import?${params.toString()}`;
}

/**
 * Build everything the dialog needs for one provider/app/model combination.
 */
export function buildConfig(options) {
  const { providerName, app, baseUrl, modelId = null, homepage = null, protocols = [], notes = null, verified = false } = options;

  const verdict = evaluate({ app, protocols, baseUrl });
  const fields = buildFields({ providerName, app, baseUrl, modelId, homepage, notes });

  const warnings = [];
  if (verdict.status === STATUS.UNKNOWN) {
    warnings.push('来源未声明协议，无法判断该服务是否满足目标工具的调用路径。已标记为 unknown，不会生成可用的导入链接。');
  }
  if (verdict.status === STATUS.UNSUPPORTED_PROTOCOL) {
    warnings.push('该服务的协议与目标应用不匹配，Radar 不会生成导入链接，避免你导入一个注定不能用的配置。');
  }
  warnings.push('该配置未经过真实 API 调用验证。Radar 只确认字段格式与协议声明，不代表密钥有效或额度可用。');
  if (!verified) {
    warnings.push('该来源的官方域名尚未核对，请自行确认地址没有被第三方转手。');
  }

  const canLink = verdict.status === STATUS.READY_WITHOUT_KEY;

  return {
    providerName,
    app,
    appLabel: APP_LABELS[app] ?? app,
    baseUrl: normaliseEndpoint(baseUrl),
    modelId,
    protocols,
    status: verdict.status,
    reason: verdict.reason,
    requiresConversion: verdict.requiresConversion,
    protocol: verdict.protocol,
    protocolLabel: verdict.protocol ? PROTOCOL_LABELS[verdict.protocol] ?? verdict.protocol : null,
    fields,
    warnings,
    deepLink: canLink ? buildDeepLink({ app, providerName, baseUrl, modelId, homepage, notes }) : null,
    supportedVersion: SUPPORTED_VERSION,
    supportedCommit: SUPPORTED_COMMIT,
    radarExportLabel: 'Radar 数据交换文件（非 CC Switch 原生导入文件）',
  };
}

/** Plain-text rendering of every copiable field, for the copy-all action. */
export function fieldsToText(fields) {
  return fields
    .filter((field) => field.value !== null && field.value !== undefined && field.value !== '')
    .map((field) => `${field.label}：${field.value}`)
    .join('\n');
}

File: radar/cc_switch.py
"""CC Switch configuration builder (``python -m radar.cc_switch``).

Builds the import payload for CC Switch, anchored to the **official V1 deep
link protocol** documented at
``docs/user-manual/zh/5-faq/5.3-deeplink.md`` in ``farion1231/cc-switch``.

Verified protocol shape (quoted from the official documentation, see
``docs/cc-switch.md`` for the full record):

    ccswitch://v1/import?resource={type}&app={app}&name={name}&...

    resource  required  provider | mcp | prompt | skill
    app       required  claude | codex | gemini | opencode | openclaw
    name      required
    endpoint  optional  API endpoint; comma-separated list is allowed
    apiKey    optional  API key
    homepage  optional
    model     optional
    ...

WHAT THIS MODULE DELIBERATELY DOES NOT DO
-----------------------------------------
* It never invents a JSON blob and claims CC Switch can import it.
* It never writes to CC Switch's SQLite database and never performs a
  full-database SQL restore.
* It never overwrites a user's existing configuration.
* It never puts an API key in a link. Radar does not collect upstream keys at
  all; the user fills their own key inside CC Switch.

Protocol fidelity
-----------------
``openai_chat``, ``openai_responses``, ``anthropic_messages`` and
``gemini_native`` are tracked separately. A service that only supports Chat
Completions is NOT reported as compatible with a path that requires Responses.
When support cannot be established the status is ``unknown`` or
``unsupported`` -- never a fabricated success.
"""

from __future__ import annotations

import argparse
import json
import re
import sys
from collections.abc import Iterable
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any
from urllib.parse import quote, urlsplit, urlunsplit

from .vocab import (
    CCApp,
    CCResource,
    CCStatus,
    Protocol,
)

#: The CC Switch release this integration was written and verified against.
#: Pinning a baseline is required so a later GUI change cannot silently break
#: the generated links.
SUPPORTED_CC_SWITCH_VERSION = "3.20.3"
SUPPORTED_CC_SWITCH_REF = "v3.20.3"
SUPPORTED_CC_SWITCH_COMMIT = "06082e1"

#: Deep link version segment confirmed by the official protocol document.
DEEP_LINK_VERSION = "v1"

#: Maximum lengths enforced as a whitelist guard before anything is emitted.
MAX_FIELD_LENGTH = 2048
MAX_NAME_LENGTH = 120

#: Characters permitted in an endpoint URL once assembled.
_ENDPOINT_PATTERN = re.compile(r"^https?://[^\s<>\"'\\]+$")

#: Which CC Switch target applications a given protocol can serve.
#:
#: ``claude`` drives Claude Code, which speaks the Anthropic Messages API.
#: ``codex`` drives Codex, which uses the OpenAI Responses API.
#: ``gemini`` drives the Gemini CLI, which uses the native Gemini API.
#: ``opencode`` and ``openclaw`` are recorded as OpenAI Chat Completions
#: clients here, which is the behaviour of the pinned release.
APP_PROTOCOL_REQUIREMENTS: dict[str, set[Protocol]] = {
    CCApp.CLAUDE.value: {Protocol.ANTHROPIC_MESSAGES},
    CCApp.CODEX.value: {Protocol.OPENAI_RESPONSES},
    CCApp.GEMINI.value: {Protocol.GEMINI_NATIVE},
    CCApp.OPENCODE.value: {Protocol.OPENAI_CHAT},
    CCApp.OPENCLAW.value: {Protocol.OPENAI_CHAT},
}

APP_DISPLAY_NAMES = {
    CCApp.CLAUDE.value: "Claude Code",
    CCApp.CODEX.value: "Codex",
    CCApp.GEMINI.value: "Gemini CLI",
    CCApp.OPENCODE.value: "OpenCode",
    CCApp.OPENCLAW.value: "OpenClaw",
}

#: Human label for each protocol, used in the UI instead of a green tick that
#: would imply blanket compatibility.
PROTOCOL_LABELS = {
    Protocol.OPENAI_CHAT.value: "Chat Completions",
    Protocol.OPENAI_RESPONSES.value: "Responses",
    Protocol.ANTHROPIC_MESSAGES.value: "Messages (Anthropic)",
    Protocol.GEMINI_NATIVE.value: "Gemini native",
}


@dataclass
class CCField:
    """One configurable field shown in the preview and copied by the user."""

    key: str
    label: str
    value: str | None
    required: bool = False
    hint: str | None = None
    secret: bool = False
    copiable: bool = True


@dataclass
class CCExtra:
    """Optional protocol-adjacent fields."""

    haiku_model: str | None = None
    sonnet_model: str | None = None
    opus_model: str | None = None


@dataclass
class CCConfig:
    """Result of building an import payload for one provider/app pair."""

    status: str
    target: str | None
    target_label: str | None
    supported_version: str
    supported_commit: str
    resource: str
    fields: list[CCField] = field(default_factory=list)
    deep_link: str | None = None
    warnings: list[str] = field(default_factory=list)
    reason: str | None = None
    requires_conversion: bool = False
    protocol: str | None = None
    protocol_label: str | None = None
    #: Named explicitly so the UI can say what it is NOT: a CC Switch native
    #: import file requires verification; this is a Radar data-exchange file.
    radar_export: dict[str, Any] = field(default_factory=dict)

    @property
    def ok(self) -> bool:
        return self.status in {CCStatus.READY.value, CCStatus.READY_WITHOUT_KEY.value}

    def to_dict(self) -> dict[str, Any]:
        return {
            "status": self.status,
            "target": self.target,
            "target_label": self.target_label,
            "supported_version": self.supported_version,
            "supported_commit": self.supported_commit,
            "resource": self.resource,
            "protocol": self.protocol,
            "protocol_label": self.protocol_label,
            "requires_conversion": self.requires_conversion,
            "reason": self.reason,
            "warnings": self.warnings,
            "deep_link": self.deep_link,
            "fields": [
                {
                    "key": item.key,
                    "label": item.label,
                    "value": item.value,
                    "required": item.required,
                    "hint": item.hint,
                    "secret": item.secret,
                    "copiable": item.copiable,
                }
                for item in self.fields
            ],
            # Labelled unambiguously as a Radar exchange file.
            "radar_export": self.radar_export,
            "radar_export_label": "Radar 数据交换文件（非 CC Switch 原生导入文件）",
        }


# ---------------------------------------------------------------------------
# Endpoint handling
# ---------------------------------------------------------------------------


def normalise_endpoint(base_url: str | None) -> str | None:
    """Validate and canonicalise an API base URL.

    The task book calls out path-joining rules explicitly: the existing ``/v1``
    must not be duplicated, a trailing slash is collapsed, and a URL that
    already carries the full endpoint path is kept as-is. Chinese characters
    and other non-ASCII in a path are percent-encoded.
    """
    if not base_url:
        return None
    text = base_url.strip().strip("`'\"")

    # A URL that already ends in a known completion path stays untouched.
    parts = urlsplit(text)
    if parts.scheme not in {"http", "https"} or not parts.hostname:
        return None

    path = parts.path or ""

    # Collapse a duplicated version segment such as /v1/v1.
    path = re.sub(r"/v1/v1(/|$)", r"/v1\1", path)
    path = re.sub(r"/v1beta/v1beta(/|$)", r"/v1beta\1", path)

    # Remove a trailing slash so the caller can join consistently.
    path = path.rstrip("/")

    # Percent-encode non-ASCII without double-encoding existing escapes.
    if any(ord(character) > 127 for character in path):
        path = quote(path, safe="/:@!$&'()*+,;=~-._")

    rebuilt = urlunsplit((parts.scheme, parts.netloc, path, parts.query, ""))
    if not _ENDPOINT_PATTERN.match(rebuilt):
        return None
    return rebuilt


def join_endpoint(base_url: str, suffix: str) -> str | None:
    """Join a path suffix without duplicating a version segment."""
    base = normalise_endpoint(base_url)
    if base is None:
        return None
    tail = "/" + suffix.strip("/")
    if base.endswith(tail):
        return base
    if "/v1" in base and tail.startswith("/v1"):
        tail = tail[3:]
    return base + tail


def is_valid_endpoint(url: str | None) -> bool:
    if not url:
        return False
    return bool(_ENDPOINT_PATTERN.match(url)) and len(url) <= MAX_FIELD_LENGTH


# ---------------------------------------------------------------------------
# Core builder
# ---------------------------------------------------------------------------


def build_config(
    *,
    provider_name: str,
    app: str,
    base_url: str | None,
    model_id: str | None = None,
    protocols: Iterable[str] = (),
    homepage: str | None = None,
    display_name: str | None = None,
    extra: CCExtra | None = None,
    notes: str | None = None,
    include_key_placeholder: bool = False,
    verified: bool = False,
) -> CCConfig:
    """Build the CC Switch import payload for one provider/app pair.

    Returns a structured result rather than raising, so the UI can always show
    a clear next step. Nothing here fetches anything or probes localhost.
    """
    protocol_set = {_as_protocol(value) for value in protocols}
    protocol_set.discard(None)  # type: ignore[arg-type]

    target = app if app in APP_PROTOCOL_REQUIREMENTS else None
    warnings: list[str] = []

    deep_link: str | None = None

    if target is None:
        return CCConfig(
            status=CCStatus.UNKNOWN.value,
            target=None,
            target_label=None,
            supported_version=SUPPORTED_CC_SWITCH_VERSION,
            supported_commit=SUPPORTED_CC_SWITCH_COMMIT,
            resource=CCResource.PROVIDER.value,
            warnings=[f"未知的目标应用：{app!r}"],
            reason="unsupported_target",
            radar_export=_radar_export(provider_name, app, base_url, model_id, protocols),
        )

    target_label = APP_DISPLAY_NAMES.get(target, target)
    required = APP_PROTOCOL_REQUIREMENTS[target]
    matched = protocol_set & required

    field_protocol = next(iter(matched)) if matched else None

    fields: list[CCField] = []
    endpoint = normalise_endpoint(base_url)

    # -- protocol gate ----------------------------------------------------
    if not protocol_set:
        status = CCStatus.UNKNOWN.value
        reason = "protocol_unknown"
        warnings.append(
            "来源未声明协议，无法判断该服务是否满足目标工具的调用路径。"
            "已标记为 unknown，不会生成可用的导入链接。"
        )
    elif not matched:
        status = CCStatus.UNSUPPORTED_PROTOCOL.value
        reason = "protocol_not_supported"
        labels = "、".join(
            PROTOCOL_LABELS.get(value.value if hasattr(value, "value") else str(value), str(value))
            for value in sorted(protocol_set, key=lambda p: str(p))
        )
        warnings.append(
            f"该服务声明支持 {labels}，但 {target_label} 需要 "
            f"{'、'.join(PROTOCOL_LABELS.get(p.value, p.value) for p in sorted(required, key=lambda x: x.value))}。"
            "仅支持 Chat Completions 的服务不能直接用于需要 Responses 的调用路径。"
        )
    elif endpoint is None:
        status = CCStatus.MISSING_FIELDS.value
        reason = "missing_endpoint"
        warnings.append("缺少可用的 Base URL，无法生成导入链接。")
    elif not is_valid_endpoint(endpoint):
        status = CCStatus.MISSING_FIELDS.value
        reason = "invalid_endpoint"
        warnings.append("Base URL 未通过格式白名单校验，已拒绝生成链接。")
    else:
        status = CCStatus.READY_WITHOUT_KEY.value
        reason = "ok"

    # -- fields -----------------------------------------------------------
    fields.append(
        CCField(
            key="name",
            label="Provider 名称",
            value=_safe_text(display_name or provider_name, MAX_NAME_LENGTH),
            required=True,
            hint="在 CC Switch 中显示的供应商名称",
        )
    )
    fields.append(
        CCField(
            key="app",
            label="目标应用",
            value=target,
            required=True,
            hint=target_label,
            copiable=False,
        )
    )
    fields.append(
        CCField(
            key="endpoint",
            label="Base URL",
            value=endpoint,
            required=True,
            hint="API 端点地址；Radar 已检查 /v1 拼接规则，不会重复拼接",
        )
    )
    fields.append(
        CCField(
            key="model",
            label="模型 ID",
            value=_safe_text(model_id, 200) if model_id else None,
            required=False,
            hint="精确的上游模型标识",
        )
    )
    fields.append(
        CCField(
            key="homepage",
            label="供应商官网",
            value=_safe_text(homepage, MAX_FIELD_LENGTH) if homepage else None,
            required=False,
            hint="可选",
        )
    )
    if field_protocol is not None:
        fields.append(
            CCField(
                key="protocol",
                label="协议",
                value=PROTOCOL_LABELS.get(field_protocol.value, field_protocol.value),
                required=False,
                hint="以文本明确标注，不使用一个勾号表示全部兼容",
                copiable=True,
            )
        )

    if extra is not None and target == CCApp.CLAUDE.value:
        for key, label, value in (
            ("haikuModel", "Haiku 模型", extra.haiku_model),
            ("sonnetModel", "Sonnet 模型", extra.sonnet_model),
            ("opusModel", "Opus 模型", extra.opus_model),
        ):
            if value:
                fields.append(
                    CCField(
                        key=key,
                        label=label,
                        value=_safe_text(value, 200),
                        required=False,
                        hint="仅 Claude 支持",
                    )
                )

    if notes:
        fields.append(
            CCField(
                key="notes",
                label="备注",
                value=_safe_text(notes, 400),
                required=False,
                hint="会写入 CC Switch 的备注字段",
            )
        )

    # The user's own key is never collected by Radar. The field is shown so the
    # user knows what to paste inside CC Switch, and is marked as a secret so
    # it never enters a URL or an export.
    fields.append(
        CCField(
            key="apiKey",
            label="API Key",
            value=None,
            required=True,
            hint=("请在你自己的 CC Switch 中填写。Radar 不收集、不传输、" "不记录任何上游密钥。"),
            secret=True,
            copiable=False,
        )
    )

    # -- deep link --------------------------------------------------------
    if status in {CCStatus.READY.value, CCStatus.READY_WITHOUT_KEY.value}:
        deep_link = build_deep_link(
            resource=CCResource.PROVIDER.value,
            app=target,
            name=display_name or provider_name,
            endpoint=endpoint,
            homepage=_safe_text(homepage, MAX_FIELD_LENGTH) if homepage else None,
            model=_safe_text(model_id, 200) if model_id else None,
            api_key=None,  # never populated by Radar
            include_key_placeholder=include_key_placeholder,
        )
        if include_key_placeholder:
            warnings.append(
                "链接中包含 API Key 占位符。CC Switch 会在导入确认框中显示该字段，"
                "请在导入后替换为你自己的密钥。"
            )

    if not verified:
        warnings.append(
            "该配置未经过真实 API 调用验证。Radar 只确认字段格式与协议声明，"
            "不代表密钥有效或额度可用。"
        )

    return CCConfig(
        status=status,
        target=target,
        target_label=target_label,
        supported_version=SUPPORTED_CC_SWITCH_VERSION,
        supported_commit=SUPPORTED_CC_SWITCH_COMMIT,
        resource=CCResource.PROVIDER.value,
        fields=fields,
        deep_link=deep_link,
        warnings=warnings,
        reason=reason,
        requires_conversion=_requires_conversion(protocol_set, required),
        protocol=field_protocol.value if field_protocol else None,
        protocol_label=(PROTOCOL_LABELS.get(field_protocol.value) if field_protocol else None),
        radar_export=_radar_export(provider_name, app, base_url, model_id, protocols),
    )


def _requires_conversion(available: set[Any], required: set[Protocol]) -> bool:
    """Whether a protocol translation step would be needed.

    Only true when the service does have an OpenAI-compatible path but the
    target needs a different one. It is never guessed beyond what the sources
    actually declare.
    """
    if not available or not required:
        return False
    if available & required:
        return False
    openai_family = {Protocol.OPENAI_CHAT, Protocol.OPENAI_RESPONSES}
    return bool(available & openai_family) and bool(required & openai_family)


def build_deep_link(
    *,
    resource: str,
    app: str,
    name: str,
    endpoint: str | None = None,
    homepage: str | None = None,
    model: str | None = None,
    api_key: str | None = None,
    include_key_placeholder: bool = False,
) -> str:
    """Assemble a ``ccswitch://v1/import`` URL exactly per the V1 protocol.

    ``urllib``-style percent encoding is applied to every parameter value, so
    Chinese names and special characters are encoded correctly.
    """
    from urllib.parse import urlencode

    params: list[tuple[str, str]] = [
        ("resource", resource),
        ("app", app),
        ("name", name),
    ]
    if endpoint:
        params.append(("endpoint", endpoint))
    if homepage:
        params.append(("homepage", homepage))
    if model:
        params.append(("model", model))

    if api_key:
        # Radar never does this with a real key. The branch exists solely so a
        # future integration could, and it is guarded by the caller.
        params.append(("apiKey", api_key))
    elif include_key_placeholder:
        params.append(("apiKey", "YOUR_API_KEY"))

    query = urlencode(params, quote_via=quote, safe="")
    return f"ccswitch://{DEEP_LINK_VERSION}/import?{query}"


def _radar_export(
    provider_name: str,
    app: str,
    base_url: str | None,
    model_id: str | None,
    protocols: Iterable[str],
) -> dict[str, Any]:
    """The project's own JSON export.

    Named and described as a Radar data-exchange file. It must NOT be called a
    CC Switch native import file unless that has actually been verified.
    """
    return {
        "kind": "freeai-radar.exchange.v1",
        "notice": (
            "这是 FreeAI Radar 的数据交换文件，不是 CC Switch 原生导入文件。"
            "CC Switch 原生的批量导入能力请使用 ccswitch:// 深度链接。"
        ),
        "provider_name": provider_name,
        "target_app": app,
        "base_url": normalise_endpoint(base_url),
        "model_id": model_id,
        "protocols": sorted(
            str(value.value if hasattr(value, "value") else value) for value in protocols
        ),
        "api_key": None,
    }


def _as_protocol(value: Any) -> Protocol | None:
    if isinstance(value, Protocol):
        return value
    try:
        return Protocol(str(value))
    except ValueError:
        return None


def _safe_text(value: str | None, limit: int) -> str | None:
    """Whitelist-guard free text before it enters a link or the DOM.

    Control characters and angle brackets are removed, and the length is
    capped. Scraped content can therefore never inject markup or script.
    """
    if value is None:
        return None
    text = re.sub(r"[\x00-\x1f\x7f<>]", "", str(value)).strip()
    if not text:
        return None
    if len(text) > limit:
        text = text[: limit - 1].rstrip() + "…"
    return text


# ---------------------------------------------------------------------------
# Batch build for the static export
# ---------------------------------------------------------------------------


def build_all(
    catalog: dict[str, Any],
    *,
    apps: Iterable[str] = (CCApp.CLAUDE.value, CCApp.CODEX.value, CCApp.GEMINI.value),
) -> dict[str, Any]:
    """Pre-compute CC Switch payloads for every provider in the catalog.

    Emitting this at build time keeps the browser work trivial and means the
    same logic is exercised by the tests. The browser also re-runs the pure
    function when a user changes the target tool.
    """
    providers = {provider["id"]: provider for provider in catalog.get("providers", [])}
    models_by_provider: dict[str, list[dict[str, Any]]] = {}
    for model in catalog.get("models", []):
        models_by_provider.setdefault(model["provider_id"], []).append(model)

    offers_by_provider: dict[str, list[dict[str, Any]]] = {}
    for offer in catalog.get("offers", []):
        offers_by_provider.setdefault(offer["provider_id"], []).append(offer)

    payload: dict[str, Any] = {
        "schema_version": 1,
        "supported_version": SUPPORTED_CC_SWITCH_VERSION,
        "supported_ref": SUPPORTED_CC_SWITCH_REF,
        "supported_commit": SUPPORTED_CC_SWITCH_COMMIT,
        "protocol": f"ccswitch://{DEEP_LINK_VERSION}/import",
        "entries": {},
    }

    for provider_id, provider in providers.items():
        models = models_by_provider.get(provider_id, [])
        offers = offers_by_provider.get(provider_id, [])

        base_url = next((offer.get("base_url") for offer in offers if offer.get("base_url")), None)
        protocols: set[str] = set()
        for model in models:
            protocols.update(model.get("protocols") or [])
        primary_model = next((model for model in models if model.get("model_id")), None)

        entries: dict[str, Any] = {}
        for app in apps:
            config = build_config(
                provider_name=provider.get("name") or provider_id,
                display_name=provider.get("name"),
                app=app,
                base_url=base_url,
                model_id=(primary_model or {}).get("model_id"),
                protocols=protocols,
                homepage=provider.get("homepage_url"),
                verified=False,
            )
            entries[app] = config.to_dict()

        payload["entries"][provider_id] = {
            "provider_slug": provider.get("slug"),
            "provider_name": provider.get("name"),
            "base_url": base_url,
            "protocols": sorted(protocols),
            "apps": entries,
        }

    return payload


def parse_args(argv: list[str] | None = None) -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        prog="python -m radar.cc_switch",
        description="Pre-compute CC Switch import payloads from a public catalog.",
    )
    parser.add_argument("--catalog", required=True, help="path to catalog.<version>.json")
    parser.add_argument("--output", required=True, help="output JSON path")
    parser.add_argument(
        "--apps",
        nargs="*",
        default=[CCApp.CLAUDE.value, CCApp.CODEX.value, CCApp.GEMINI.value],
    )
    return parser.parse_args(argv)


def main(argv: list[str] | None = None) -> int:
    args = parse_args(argv)
    catalog = json.loads(Path(args.catalog).read_text(encoding="utf-8"))
    payload = build_all(catalog, apps=args.apps)

    output = Path(args.output)
    output.parent.mkdir(parents=True, exist_ok=True)
    output.write_text(json.dumps(payload, ensure_ascii=False, indent=2) + "\n", encoding="utf-8")

    statuses: dict[str, int] = {}
    for entry in payload["entries"].values():
        for app_config in entry["apps"].values():
            status = app_config["status"]
            statuses[status] = statuses.get(status, 0) + 1

    print(
        json.dumps(
            {
                "providers": len(payload["entries"]),
                "statuses": statuses,
                "supported_version": payload["supported_version"],
                "output": str(output),
            },
            ensure_ascii=False,
            indent=2,
        )
    )
    return 0


if __name__ == "__main__":
    sys.exit(main())
